app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -5,24 +5,43 @@ class OmniauthCallbacksController < ApplicationController
   skip_before_action :authenticate_dsi_user!
   skip_before_action :handle_expired_session!
 
-  def dfe
-    auth = request.env["omniauth.auth"]
-
-    unless DfESignIn.bypass?
-      role = DfESignInApi::GetUserAccessToService.new(
-        org_id: auth.extra.raw_info.organisation.id,
-        user_id: auth.uid,
-      ).call
+  before_action :add_auth_attributes_to_session, only: :dfe
 
+  def dfe
+    if DfESignIn.bypass?
+      create_or_update_dsi_user
+    else
+      role = check_user_access_to_service
       return redirect_to not_authorised_path unless role
+
+      create_or_update_dsi_user(role)
     end
 
-    @dsi_user = DsiUser.create_or_update_from_dsi(auth, role)
+    redirect_to root_path
+  end
+  alias_method :dfe_bypass, :dfe
+
+  private
 
+  def auth
+    request.env["omniauth.auth"]
+  end
+
+  def add_auth_attributes_to_session
+    session[:id_token] = auth.credentials.id_token
+    session[:organisation_name] = auth.extra.raw_info.organisation.name
+  end
+
+  def create_or_update_dsi_user(role = nil)
+    @dsi_user = DsiUser.create_or_update_from_dsi(auth, role)
     session[:dsi_user_id] = @dsi_user.id
     session[:dsi_user_session_expiry] = 2.hours.from_now.to_i
+  end
 
-    redirect_to root_path
+  def check_user_access_to_service
```

Comment:
@felixclack [made a point on a related PR](https://github.com/DFE-Digital/access-your-teaching-qualifications/pull/408#discussion_r1381384124) that renaming this method to `role` simplifies the code somewhat. Not a blocker.